route-rs-runtime/src/link/sync_link.rs

```diff
@@ -1,22 +1,73 @@
 use crate::element::Element;
-use crate::link::PacketStream;
+use crate::link::{ElementLinkBuilder, Link, LinkBuilder, PacketStream};
 use futures::{Async, Poll, Stream};
 
-pub struct SyncLink<E: Element> {
-    input_stream: PacketStream<E::Input>,
+#[derive(Default)]
+pub struct SyncLinkBuilder<E: Element> {
+    in_streams: Option<Vec<PacketStream<E::Input>>>,
+    element: Option<E>,
+}
+
+impl<E: Element> SyncLinkBuilder<E> {
+    pub fn new() -> Self {
+        SyncLinkBuilder {
+            in_streams: None,
+            element: None,
+        }
+    }
+}
+
+/// Although `Link` allows an arbitrary number of ingressors and egressors, `SyncLink`
+/// may only have one ingress and egress stream since it lacks some kind of queue
+/// storage. In the future we might decide to restrict the interface for this link
+/// for clearer intent.
+impl<E: Element + Send + 'static> LinkBuilder<E::Input, E::Output> for SyncLinkBuilder<E> {
+    fn ingressors(self, in_streams: Vec<PacketStream<E::Input>>) -> Self {
+        assert_eq!(in_streams.len(), 1, "SyncLink may only take 1 input stream");
+
+        SyncLinkBuilder {
+            in_streams: Some(in_streams),
+            element: self.element,
+        }
+    }
+
+    fn build_link(self) -> Link<E::Output> {
+        if self.in_streams.is_none() {
+            panic!("Cannot build link! Missing input streams");
+        } else if self.element.is_none() {
+            panic!("Cannot build link! Missing element");
+        } else {
+            let processor = SyncProcessor::new(self.in_streams.unwrap(), self.element.unwrap());
+            (vec![], vec![Box::new(processor)])
+        }
+    }
+}
+
+impl<E: Element + Send + 'static> ElementLinkBuilder<E> for SyncLinkBuilder<E> {
+    fn element(self, element: E) -> Self {
+        SyncLinkBuilder {
+            in_streams: self.in_streams,
+            element: Some(element),
+        }
+    }
+}
+
+/// The single egressor of SyncLink
+struct SyncProcessor<E: Element> {
```

Comment:
The idea behind "Processor" is that this is the thing that performs the transformation based on the `Link`'s `Element::process`, rather than just a handle for outgoing packets. It's a little strange since this entity is semantically different in different contexts: inside `SyncLink` it's a packet processor, but outside `SyncLink` it's an upstream `PacketStream` handle.

Reply:
I like the processor name, it tells us what it does in the code. And to the user, they never see the name. They only really interact with it through the TokioRunnables and Egressors. Seems like a good name to me.